### PR TITLE
Preserve Query in nextUrl during openid login redirect

### PR DESCRIPTION
**Pull request by KleinSebastian**
### Description
Successful Openid Login redirects back to the URL, from where the login was initiated. 

### Category
Bug fix 

### Why these changes are required?

Bug was reported: https://github.com/opensearch-project/security-dashboards-plugin/issues/1823 and https://github.com/opensearch-project/security-dashboards-plugin/issues/2119

### What is the old behaviour before changes and new behaviour after changes?
The changed behaviour is best explained by following user-flow
* User is not logged in
* User clicks on a shared Dashboard Link 
* User is redirected to configured Openid IDP
* After successful login, user is redirected to
  * Before:  /app/dashboards 
  * After: /app/dashboards#/view/....
* Optional: When Multitenancy is enabled, the tenant from the link was discarded before and is now kept.

The whole functionality is basically ported from the SAML login method.

### Issues Resolved
Fix: https://github.com/opensearch-project/security-dashboards-plugin/issues/1823
Fix: https://github.com/opensearch-project/security-dashboards-plugin/issues/2119

### Testing
Existing Tests pass

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
@KleinSebastian Can you add a test to go along with the changes being introduced in this PR?

**Comment by KleinSebastian** [comment]
@cwperks I added a set of unit tests for the handleUnauthRequest call. E2E Tests for the functionality are unfortunately not working on my machine, so I skipped them. If you insist upon Cypress Tests I will try my best to do it on another machine.

File: release-notes/opensearch-security-dashboards-plugin.release-notes-2.18.0.0.md
Comment (cwperks) [inline release-notes/opensearch-security-dashboards-plugin.release-notes-2.18.0.0.md:1]: Looks like this was inadvertently added.

**Comment by cwperks** [comment]
@KleinSebastian There are a few test errors. Are these errors due to changes from this PR or was there a change in OSD that introduced these errors?

**Comment by KleinSebastian** [comment]
@cwperks I was able to reproduce the error. It was an issue with updating the "security_authentication" cookie due to the mismatch of "localhost" and "127.0.0.1" between OSD and Keycloak, resulting rejection of set-cookie in the browser due to the default SameSite behaviour "Lax". 

Fixing the configuration does fix the E2E Tests as well as the need for a second redirect the existing E2E Tests.

**Comment by cwperks** [comment]
> @cwperks I was able to reproduce the error. It was an issue with updating the "security_authentication" cookie due to the mismatch of "localhost" and "127.0.0.1" between OSD and Keycloak, resulting rejection of set-cookie in the browser due to the default SameSite behaviour "Lax".
> 
> Fixing the configuration does fix the E2E Tests as well as the need for a second redirect the existing E2E Tests.

Awesome. Thank you. I just approved the CI checks to run again. We will get results in ~45min - 1 hr

**Comment by KleinSebastian** [comment]
@cwperks I saw that the E2E test runs failed again, but this time the OSD Server did not even start properly. I also saw that I introduced unwanted line ending changes. I reverted them because this might be the reason for the not starting server.

**Comment by cwperks** [comment]
> @cwperks I saw that the E2E test runs failed again, but this time the OSD Server did not even start properly. I also saw that I introduced unwanted line ending changes. I reverted them because this might be the reason for the not starting server.

I just started the CI checks again.

**Comment by KleinSebastian** [comment]
OSD is still not starting, I assume that the server is not able to contact Keycloak by using the "127.0.0.1" address and exits. I reverted the changes done to the E2E Setup and tried fixing the tests with the manual redirect workaround present in the existing tests.

File: server/auth/types/openid/openid_auth.ts
Comment (cwperks) [inline server/auth/types/openid/openid_auth.ts:136]: Thank you @KleinSebastian. I started the CI checks again. This PR looks good to me and has extensive tests. Thank you for adding those?

Quick question: https://github.com/opensearch-project/security-dashboards-plugin/issues/2119 also reports that the `#hash` part of a url is getting dropped. I see that this fix will carry url params over, but will it also carry the hash if there is one?

Comment (KleinSebastian) [reply server/auth/types/openid/openid_auth.ts:136]: Carrying over all the data after the hash was already implemented in a previous PR, but was bugged . With the small change in https://github.com/opensearch-project/security-dashboards-plugin/pull/2140/files/d5bb3126b3610f3264b112d4b8adb12a8f18dcd7#diff-55f93c47c667faee3f68b8b8f5e98157fb36eefdc7b856355d5a9b4350526070R104 the functionality of capturing and restoring the hash part is working as expected.

Comment (cwperks) [reply server/auth/types/openid/openid_auth.ts:136]: Got it. Thank you for confirming!

**Review by cwperks — APPROVED** [review]

**Comment by KleinSebastian** [comment]
@cwperks Are there some additional actions required from my side? Its my first time performing a PR on a public repo, so I'm a bit unsure about the process.

**Comment by cwperks** [comment]
The PR needs 2 maintainer approvals and some are out of the office this week. @DarshitChanpura or @RyanL1997 could you take a look at this PR?

**Review by DarshitChanpura — APPROVED** [review]
LGTM! Thanks @KleinSebastian !

**Review by RyanL1997 — APPROVED** [review]

**Pull request merged**
